pkgs/tools/security/pinentry/default.nix

```diff
@@ -1,60 +1,96 @@
 { fetchurl, fetchpatch, stdenv, lib, pkgconfig, autoreconfHook
 , libgpgerror, libassuan
-, libcap ? null, libsecret ? null, ncurses ? null, gtk2 ? null, gcr ? null
-, qt4 ? null, qt5 ? null
-, enableEmacs ? false
+, ncurses, gtk2, qt
+, libcap ? null, libsecret ? null, gcr ? null
+, enabledFlavors ? [ "curses" "tty" "gtk2" "qt" "gnome3" "emacs" ]
 }:
 
-assert qt5 != null -> qt4 == null;
-assert qt4 != null -> qt5 == null;
+with stdenv.lib;
+
+assert isList enabledFlavors && enabledFlavors != [];
 
 let
   mkDerivation =
-    if qt5 != null
-      then qt5.mkDerivation
+    if (builtins.elem "qt" enabledFlavors)
+      then qt.mkDerivation
       else stdenv.mkDerivation;
+
+  mkFlag = pfxTrue: pfxFalse: cond: name:
+    "--${if cond then pfxTrue else pfxFalse}-${name}";
+  mkEnable = mkFlag "enable" "disable";
+  mkWith = mkFlag "with" "without";
+
+  mkEnablePinentry = f:
+    let
+      info = flavorInfo.${f};
+      inputs = info.buildInputs or [];
+      flag = flavorInfo.${f}.flag or null;
+      inputsSatisfied = all (f: ! isNull f) inputs;
+    in
+      assert inputsSatisfied;
+      optionalString (flag != null)
+        (mkEnable (elem f enabledFlavors && inputsSatisfied) ("pinentry-" + flag));
+
+  flavorInfo = {
+    curses = { bin = "curses"; buildInputs = [ ncurses ]; };
```

Comment:
Hmm, what happens if `curses` is not in `enabledFlavors`

Reply:
`pinentry` is built without ncurses support, so `pinentry-tty` is used?

Reply:
Hmm, it enables it based on the [presence of the library](https://github.com/GPGTools/pinentry/blob/b75069d5bd129d58c3976aa73726773317a69fab/configure.ac#L313-L315). Maybe we should also add `--disable-{pinentry,fallback}-curses` to avoid building it when it is not in `enabledFlavors` but is propagated into inputs.

Reply:
Is there some tooling to generate these `--enable`/`--disable` statements?

I don't have a strong opinion about whether to ship pinentry with or without ncurses. Do we need a pinentry in some non-interactive gpg workloads, where avoiding pulling in ncurses would be desireable?

Reply:
I added `flag = "curses"` here, so we'll automatically configure a `--disable-pinentry-curses` if it's not in `enabledFlavors`.